dex/wait/queue_test.go

```diff
@@ -63,41 +64,45 @@ func TestTaper(t *testing.T) {
 }
 
 func TestTaperingQueue(t *testing.T) {
-	q := NewTaperingTickerQueue(time.Millisecond, time.Millisecond*5)
-
-	var waiterNumber int
-	var resultMtx sync.Mutex
-	var resultOrder []int
-	var wg sync.WaitGroup
-	addWaiter := func(numTryAgains int) {
+	const fastestInterval, slowestInterval = 1 * time.Millisecond, 2 * time.Millisecond
+	expiration := time.Now().Add(time.Minute)
+
+	q := NewTaperingTickerQueue(fastestInterval, slowestInterval)
+
+	// waiterTriesTimedMtx protects waiterTriesTimed from concurrent access.
+	var waiterTriesTimedMtx sync.Mutex
+	// waiterTriesTimed maps waiter to a list of tries, each try is represented
+	// by timestamp (that reflects when waiter try starts executing).
+	waiterTriesTimed := make(map[int][]time.Time, 5)
+	var wgWaiters sync.WaitGroup
+	addWaiter := func(waiterNumber, numTryAgains int) {
 		var numTrys int
-		num := waiterNumber
-		waiterNumber++
 		q.Wait(&Waiter{
-			Expiration: time.Now().Add(time.Hour),
+			Expiration: expiration,
 			TryFunc: func() TryDirective {
+				waiterTriesTimedMtx.Lock()
+				// Record when try func was called to check it later.
+				waiterTriesTimed[waiterNumber] = append(waiterTriesTimed[waiterNumber], time.Now())
+				waiterTriesTimedMtx.Unlock()
 				numTrys++
 				if numTrys > numTryAgains {
-					resultMtx.Lock()
-					resultOrder = append(resultOrder, num)
-					resultMtx.Unlock()
-					wg.Done()
+					wgWaiters.Done()
 					return DontTryAgain
 				}
 				return TryAgain
 			},
-			ExpireFunc: func() {},
+			// We don't expect expire func being called in this test, leaving it
+			// undefined so that we'll get a panic in case it gets called.
+			//ExpireFunc: func() {},
```

Comment:
maybe worth covering as well, maybe in a separate test ?
- on one hand, this test is already quite complex
- on the other, when testing concurrency its better to combine lots of different go-routines executing different code-paths

Reply:
Added simple check to `TestTaper` test.